### PR TITLE
Functional users list overlays

### LINKED ISSUE #468: Build redesigned permission screens
Builds designs decided in #441 
[Figma link](https://www.figma.com/design/DNbFdd7YrA2MpgeXNgmNLz/Cloud.gov-%E2%80%94-Dashboard-UI-redesign?node-id=870-12972&t=BXk901uIw5EiOUIS-1)

For animations, refer to [Ethan's comments](https://www.figma.com/design/DNbFdd7YrA2MpgeXNgmNLz?node-id=871-22181#934992639)

For the Overlay, first try building a new component using `<dialog'>` then go back and refactor our USWDS Modal component based on whether that meets our needs.

### Acceptance Criteria (tk)
- [x] Overlay should scroll / scrolling outside overlay is prevented
- [ ] Overlay should have same a11y requirements as modal (trap focus, escape to close, clicking outside overlay should close, etc)
- [ ] Stretch goal: link to particular user in overlay is sharable (i.e. has proper ids in url)

### PR DESCRIPTION
**Pull request by echappen**
Blocked by PR #476 

## Changes proposed in this pull request:

- Converts roles functionality in users list to overlays

Per [engineering sync discussion](https://github.com/cloud-gov/cg-ui/pull/476#issuecomment-2341945567), some additional acceptance criteria:

- [x] When opening the dialog, we'll keep focus on the close button and allow users to tab through the rest of the content.
- [x] Opening the dialog should also announce a title specific to the functionality being presented in the dialog.
- [x] When closing the dialog, we'll keep the default behavior of returning focus to the last point of regard.
- [ ] When closing the dialog, we'll have SR's announce any success messages (while keeping default focus).
- [x] While dialog is open, we agreed that a user should be able to click outside of it to close it. Since this is not default behavior for dialogs, we're going to spike on this and see what we come up with.

### How to test

1. Log into the CloudFoundry dev environment through the CLI: `cf login -a [url] --sso`
2. Start up a dev server with a refreshed CF token: `npm run dev-cf`
3. Navigate to the users list: `/orgs/[orgId]`
4. For the org roles form, click any link under the "organization roles" column
5. For the space roles form, click any link under the "access permissions" column

Forms should operate the same as when they were on pages, with the only difference being that the overlay closes on form submission success. 

### Related issues

Closes #468

### Submitter checklist

- [ ] Added logging is not capturing sensitive data and is set to an appropriate level (DEBUG vs INFO etc)
- [ ] Updated relevant documentation (README, ADRs, explainers, diagrams)

## Security considerations

None, UI changes only

### CODE REVIEW

**Comment by echappen** [comment]
@beepdotgov @cannandev @hursey013 The only criteria I'm still having trouble with is announcing the success message—can't get it to announce. I'll keep working on that, but otherwise, I think this is ready for eyes!

**Comment by echappen** [comment]
I'm going to make dynamic url params a separate GH issue—I might get to it depending on how much I need to update with this PR, but I don't want it to block the rest of the work.

File: src/components/Overlays/OverlayHeaderUsername.tsx
Comment (beepdotgov) [inline src/components/Overlays/OverlayHeaderUsername.tsx:17]: Could we bump the font size down on this slightly? (It looks to be ~16px in the prototype; I was aiming for 14px in the design.)

File: src/components/Overlays/OverlayHeaderUsername.tsx
Comment (beepdotgov) [inline src/components/Overlays/OverlayHeaderUsername.tsx:18]: [non-blocking comment, TIL] I actually didn’t know about `text-underline-offset`! (I would’ve just used border-bottom, and offset it with a little padding.)

**Review by beepdotgov — CHANGES_REQUESTED** [review]
@echappen Seriously, this is looking _fantastic_ so far. I’ve left a few design suggestions/requests, and a couple non-blocking comments. Please let me know if anything’s unclear, or if you’d like to discuss anything further!

File: src/components/Overlays/OrgUserOrgRolesOverlay.tsx
Comment (echappen) [reply src/components/Overlays/OrgUserOrgRolesOverlay.tsx:30]: done!

File: src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx
Comment (echappen) [reply src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx:154]: Thanks for pointing out—we're simply checking to see if the form has been fiddled with (on the first `handleChange` event trigger) in order to enable the button. Checking against initial state of the form on each change was much harder to implement. It won't hurt to submit the form with no changes, but I agree that the visual cue that the form has changed is a nice one. Do you think this current state is confusing enough to warrant a more sophisticated solution?

File: src/components/OverlayDrawer.tsx
Comment (echappen) [reply src/components/OverlayDrawer.tsx:7]: Would that require an additional loading indicator on the page, in case it takes a while for the drawer to open after clicking the button? (FWIW I don't expect loading to take nearly as long in prod as it does in dev)

File: src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx
Comment (beepdotgov) [reply src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx:154]: > Do you think this current state is confusing enough to warrant a more sophisticated solution?

Oh no, sorry, absolutely not! It just jumped out at me, so I thought I’d log it.

File: src/components/Overlays/OverlayHeaderUsername.tsx
Comment (echappen) [reply src/components/Overlays/OverlayHeaderUsername.tsx:18]: Me neither until recently! It seemed to have [good enough browser support](https://developer.mozilla.org/en-US/docs/Web/CSS/text-underline-offset#browser_compatibility) (if you're not using percentage values) so I thought I'd use it.

File: src/components/OverlayDrawer.tsx
Comment (beepdotgov) [reply src/components/OverlayDrawer.tsx:7]: Hmm, yeah, if the delays are going to be longish on dev and/or variable in length, I think a loading state would make sense. I’d be happy to whip one up, if that’d be helpful.

Comment (beepdotgov) [reply src/components/OverlayDrawer.tsx:7]: (but/and I could move that work over to #322 so you’re not blocked)

File: src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx
Comment (echappen) [reply src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx:154]: No apologies needed! It's never bothered me enough to take the time for a better solution, but I didn't know if others had the same read.

File: src/components/OverlayDrawer.tsx
Comment (beepdotgov) [reply src/components/OverlayDrawer.tsx:7]: Sounds great! On it.

Comment (beepdotgov) [reply src/components/OverlayDrawer.tsx:7]: Just opened #487! Resolving this for now.

File: src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx
Comment (echappen) [reply src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx:144]: I think I got a workable solution here... do you think they even need to linearize on mobile? (It didn't seem necessary to me.)

Comment (beepdotgov) [reply src/components/UsersActions/UsersActionsSpaceRoles/UsersActionsSpaceRoles.tsx:144]: This looks _great_ to me! (And no, as long as there’s room on smaller viewports, I don’t think we need to linearize.)

File: src/components/UsersList/UsersList.tsx
Comment (echappen) [reply src/components/UsersList/UsersList.tsx:132]: Hmm I seem to be getting inconsistent behavior with this one (Safari only). A few hours ago, I saw this happen as well, but now it's returning to the button as expected. I couldn't find any literature suggesting that Safari should behave differently here, but if I had to guess, maybe the `tabindex="0"` on the table container (needed for the scrollable table) might be throwing it off.

Comment (echappen) [reply src/components/UsersList/UsersList.tsx:132]: @cannandev @hursey013 Let me know if you see this happening as well

Comment (beepdotgov) [reply src/components/UsersList/UsersList.tsx:132]: I think I might’ve figured out how to reproduce this fairly reliably, at least over here:

1. If I open an overlay via the keyboard, focus is returned correctly to the opening element.
2. If I open an overlay with a mouse/pointer, focus returns to the table.

Are you seeing that on your end?

Comment (echappen) [reply src/components/UsersList/UsersList.tsx:132]: Interesting—I'm seeing it now. I think my `tabindex="0"` suspicion is correct: when I remove that from the table container and open with a click, focus just returns to the very top of the page when closing out. (Ugh)

My guess is that only sighted users will be using the mouse to click on buttons, so this may not be a huge issue for them. For those that click buttons through the keyboard, focus will be returned as expected. But this difference in behavior between browsers is indeed frustrating.

Comment (beepdotgov) [reply src/components/UsersList/UsersList.tsx:132]: Woof, well, glad you’re also able to reproduce it!

I’d love to hear what you/Claire/Brian think, but I’d be inclined to think of this as a decently-sized a11y bug? Some people may use a screen reader and a mouse, possibly in conjunction with some screen magnification software. ([VoiceOver](https://support.apple.com/guide/voiceover/use-cursor-tracking-vo15534/mac) and [NVDA](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#NavigatingWithTheMouse) both support cursor tracking.) I don’t think this needs to block this PR, necessarily, but I might suggest logging this as a UX issue.

Comment (echappen) [reply src/components/UsersList/UsersList.tsx:132]: All good points, and I'm very willing to defer to others here.

**Comment by echappen** [comment]
I’ve been working on getting aria-live regions to be more consistently read by VoiceOver. The lack of consistency still perplexes me. 

I know that in order for aria-live regions to be read, two things need to be true ([this article](https://tetralogical.com/blog/2024/05/01/why-are-my-live-regions-not-working/) is a good summary):
1. aria-live regions need to be present in the DOM on initial page load, and
2. Screen readers will only read updates to these regions (it won’t read text already present on page load)

I suspect the inconsistency is coming down to the timing between a react component re-rendering (thus re-loading that part of the DOM) and a message updating. I think the reason that the success message wasn’t being read was because these things were happening at the same time. 

Putting a 1/2 second delay on the message update [solved this issue](https://github.com/cloud-gov/cg-ui/pull/483/commits/0bdf3442148442abb3b0fbdd3a92dc2b3478b019). But I’ve also been looking into alternatives, and I’m intrigued by [this approach](https://almerosteyn.com/2017/09/aria-live-regions-in-react), which is now [an npm package](https://www.npmjs.com/package/react-aria-live). This approach uses the React [Context API](https://react.dev/reference/react/createContext), which I haven’t used at all, so I’m less comfortable with it. @cannandev @hursey013 Any thoughts on this approach and if we should try it?

File: src/components/UsersList/UsersList.tsx
Comment (cannandev) [reply src/components/UsersList/UsersList.tsx:132]: I agree with @beepdotgov. The issue doesn't need to block this PR, but let's log as an important UX issue.

**Review by cannandev — APPROVED** [review]

File: src/components/UsersList/UsersList.tsx
Comment (echappen) [reply src/components/UsersList/UsersList.tsx:132]: Logged in #496!

**Review by beepdotgov — APPROVED** [review]

**Pull request merged**
